core/trino-main/src/main/java/io/trino/execution/QueryTracker.java

```diff
@@ -199,18 +199,18 @@ private void enforceTimeLimits()
      */
     private void pruneExpiredQueries()
     {
-        if (expirationQueue.size() <= maxQueryHistory) {
+        int nonPrunedExpirationQueueSize = expirationQueue.size();
```

Comment:
IIUC, this seems like a change in behavior. Existing code also takes into account queries that have been added while the loop is running - and keeps pruning until the "ultimate" state only has history for the number `maxQueryHistory`. If that's not the intention - then please revert it. 

In the PR code - this is left to do for the next round. IMO in high qps with low latency scenario - this may have some impact. @dain do you remember if using the latest `expirationQueue.size()` was intentional here?

Reply:
I'm talking about the following scenario:

```
Say maxQueryHistory=10.
expirationQueue size initially = 15.
q1, q2, q3, q4, q5, .... q15

for (T query : expirationQueue) {
            // first iteration => count= 0, 15 <= 10?, q1 pruned, count = 1
            // second iteration => count=1, 15 - 1 = 14 <= 10?, q2 pruned, count=2
            // third iteration => count=2, 15 - 2 = 13 <= 10? q3 pruned, count=3
            // ---- Say 6 more queries expire - so the expirationqueue.size() updates: q16...q21
            // 4th iteration = count=3, 21 - 3 = 18 <= 10? q4 pruned, count=4
           // ...
          // and so on .... now since you consider 21 - you prune more. But yeah, you'll stop at q15. 

            if (expirationQueue.size() - count <= maxQueryHistory) {
                break;
            }
            query.pruneInfo();
            count++;
        }
```

Reply:
Since this is periodic pruning, it may be okay - but would like to hear @dain's thoughts here - if this was intentional.

Reply:
ok. since this is not a main purpose of this PR. let me revert it.